qemu/tests/usb_redir.py

```diff
@@ -0,0 +1,239 @@
+        usbredir_dev.set_param('id', 'usb-%s' % usbredirdev_name)
+        usbredir_dev.set_param('chardev', chardev_id)
+        usbredir_dev.set_param('filter', usbredir_filter)
+        usbredir_dev.set_param('bootindex', usbredir_bootindex)
+        usbredir_dev.set_param('bus', usbredir_bus)
+        extra_params += ' '.join([chardev.cmdline(), usbredir_dev.cmdline()])
+        return extra_params
+
+    def _rv_connection_check():
+        rv_pid = process.getoutput("pidof %s" % rv_binary)
+        cmd = 'netstat -ptn | grep "^tcp.*127.0.0.1:%s.*ESTABLISHED %s.*"'
+        cmd = cmd % (spice_port, rv_pid)
+        s, o = process.getstatusoutput(cmd)
+        if s:
+            return False
+        logging.info("netstat output:\n%s", o)
+        return True
+
+    def boot_check(info):
+        """
+        boot info check
+        """
+        return re.search(info, vm.serial_console.get_stripped_output())
+
+    def _usb_dev_verify():
+        error_context.context("Check USB device in guest", logging.info)
+        if session.cmd_status(lsusb_cmd):
+            return False
+        return True
+
+    def _kill_rv_proc():
+        s, o = process.getstatusoutput("pidof %s" % rv_binary)
+        if not s:
+            process.getoutput("killall %s" % rv_binary)
+
+    def _get_usb_mount_point():
+        """ Get redirected USB stick mount point """
+        dmesg_cmd = "dmesg | grep 'Attached SCSI removable disk'"
+        s, o = session.cmd_status_output(dmesg_cmd)
+        if s:
+            test.error("Fail to get redirected USB stick in guest.")
+        dev = re.findall(r'\[(sd\w+)\]', o)[0]
+        mounts_cmd = "cat /proc/mounts | grep /dev/%s" % dev
+        s, o = session.cmd_status_output(mounts_cmd)
+        if s:
+            s, o = session.cmd_status_output('mount /dev/%s /mnt' % dev)
+            if s:
+                test.error("Fail to mount /dev/%s, output: %s" % (dev, o))
+            mp = "/mnt"
+        else:
+            mp = re.findall(r'/dev/%s\d*\s+(\S+)\s+' % dev, o)[0]
+        return mp
+
+    def _usb_stick_io(mount_point):
+        """
+        Do I/O operations on passthrough USB stick
+        """
+        error_context.context("Read and write on USB stick ", logging.info)
+        testfile = os.path.join(mount_point, 'testfile')
+        iozone_cmd = params.get("iozone_cmd",
+                                " -a -I -r 64k -s 1m -i 0 -i 1 -f %s")
+        iozone_test.run(iozone_cmd % testfile)
+
+    error_context.context("Prepare usb-redir device", logging.info)
+    usbredirdev_name = params["usbredirdev_name"]
+    usbredir_params = params.object_params(usbredirdev_name)
+    backend = usbredir_params.get('chardev_backend', 'spicevmc')
+    if backend not in ('spicevmc', 'tcp_socket'):
+        test.error("Unsupported char device backend type: %s" % backend)
+    extra_params = _generate_usb_redir_cmdline()
+    params["extra_params"] = extra_params
+    params["start_vm"] = "yes"
```

Comment:
This line is conflict with L182 , it depends if it can connect vm before vm start, please check this, thanks!